src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java

```diff
@@ -0,0 +1,355 @@
+        private static final String CONNECT_PARTITION = "__connect_partition";
+        private static final String CONNECT_OFFSET = "__connect_offset";
+
+        // External contributed builder state
+        private PrimaryKeyMode primaryKeyMode;
+        private Set<String> primaryKeyFields;
+        private SinkRecord sinkRecord;
+        private DatabaseDialect dialect;
+
+        // Internal build state
+        private List<String> keyFieldNames = new ArrayList<>();
+        private List<String> nonKeyFieldNames = new ArrayList<>();
+        private Map<String, FieldDescriptor> allFields = new LinkedHashMap<>();
+
+        public Builder withDialect(DatabaseDialect dialect) {
+            this.dialect = dialect;
+            return this;
+        }
+
+        public Builder withPrimaryKeyFields(Set<String> primaryKeyFields) {
+            this.primaryKeyFields = primaryKeyFields;
+            return this;
+        }
+
+        public Builder withPrimaryKeyMode(PrimaryKeyMode primaryKeyMode) {
+            this.primaryKeyMode = primaryKeyMode;
+            return this;
+        }
+
+        public Builder withSinkRecord(SinkRecord record) {
+            this.sinkRecord = record;
+            return this;
+        }
+
+        public SinkRecordDescriptor build() {
+            Objects.requireNonNull(primaryKeyMode, "The primary key mode must be provided.");
+            Objects.requireNonNull(sinkRecord, "The sink record must be provided.");
+
+            final boolean flattened = isFlattened(sinkRecord);
+            readSinkRecordKeyData(sinkRecord, flattened);
+            readSinkRecordNonKeyData(sinkRecord, flattened);
+
+            return new SinkRecordDescriptor(sinkRecord, sinkRecord.topic(), keyFieldNames, nonKeyFieldNames, allFields, flattened);
+        }
+
+        private boolean isFlattened(SinkRecord record) {
+            final Schema valueSchema = record.valueSchema();
+            if (valueSchema != null) {
+                for (Field field : valueSchema.fields()) {
+                    if (Schema.Type.STRUCT.equals(field.schema().type())) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+
+        private void readSinkRecordKeyData(SinkRecord record, boolean flattened) {
+            switch (primaryKeyMode) {
+                case NONE:
+                    // does nothing
+                    break;
+                case KAFKA:
+                    applyKafkaCoordinatesAsPrimaryKey();
+                    break;
+                case RECORD_KEY:
+                    applyRecordKeyAsPrimaryKey(record);
+                    break;
+                case RECORD_VALUE:
+                    applyRecordValueAsPrimaryKey(record, flattened);
+                    break;
+                default:
+                    throw new ConnectException("Unexpected primary key mode: " + primaryKeyMode);
```

Comment:
Please use `DebeziumException` in most places unless `ConnectException` is explictly required for integration with Kafka Connect.